### PR TITLE
Add debug parameter

### PR DESCRIPTION
**Pull request by Sierraffinity**

### CODE REVIEW

File: src/gbafix.c [hunk 1]
```diff
@@ -141,7 +141,7 @@ int main(int argc, char *argv[])
 	if (argc <= 1)
 	{
 		printf("GBA ROM fixer v"VER" by Dark Fader / BlackThunder / WinterMute \n");
-		printf("Syntax: gbafix <rom.gba> [-p] [-t[title]] [-g<game_code>] [-m<maker_code>] [-v<version>]\n");
+		printf("Syntax: gbafix <rom.gba> [-p] [-t[title]] [-c<game_code>] [-m<maker_code>] [-r<version>]\n");
```
Comment (WinterMute) [inline src/gbafix.c]: This should be a separate PR, it's not relevant to this change.

**Comment by WinterMute** [comment]
What problem does this fix?

I don't think changing the default is a particularly good idea. Adding an option to change this value and retaining the current default would be less likely to cause unintentional breakage elsewhere.

**Comment by Sierraffinity** [comment]
Changing this byte does not affect usage on regular GBA hardware. According to GBATEK:

>0B4h - Device type, 1 Byte. 
>Normally, this entry should be zero. With Nintendo's hardware debugger Bit 7 identifies the debugging handlers entry point and size of DACS (Debugging And Communication System) memory: Bit7=0: 9FFC000h/8MBIT DACS, Bit7=1: 9FE2000h/1MBIT DACS. The debugging handler can be enabled in 800009Ch (see above), normal cartridges do not have any memory (nor any mirrors) at these addresses though.

I'm not sure why this was ever the default in gbafix to begin with. I do understand why this was never caught however. The pret projects (pokeruby, pokeemerald) need this byte  to be the standard zero, and since that bit only matters if you have Nintendo's hardware debugger and need to change DACS info, I thought it would be a safe change to make it zero by default.

File: src/gbafix.c (hunk 1 above)
Comment (Sierraffinity) [reply src/gbafix.c]: Sorry, I noticed that the usage screen was wrong in testing my change, so I figured it would be okay to include it here as it's such a small change. I'll split it into its own.

**Comment by WinterMute** [comment]
gbatek is incorrect or at least misleading here. GBA cart space  is 32MB @ 0x08000000 and clearly those addresses are within that range. 

The bit affects where bios exception vectors jump to, even on retail units. I'll need to go dig out a bios disassembly to say for certain but I'm sure the illegal instruction vector and the FIQ vector are affected, not certain about the others offhand but I do remember experiments with the FIQ handler.

 It's also possible to modify a retail GBA to allow access to the FIQ line.
agb-fiq
 
I fail to see why the pret projects would need the byte to be zero -  do they rely on this feature in some way?

**Comment by Sierraffinity** [comment]
Ah I see, I didn't know about that. Pret needs that byte to be zero because that's how the retail games shipped and we need to create byte-identical binaries. I could see about making it an option though instead of hardcoded. Is -d good?

**Comment by Sierraffinity** [comment]
@WinterMute I've added the `-d` parameter, does that work?

**Comment by Sierraffinity** [comment]
Doing more research on GBATEK, the "device type" seemingly doesn't change anything regarding the FIQ/illegal instruction vectors unless certain bits in 0x9C are set:

> 09Ch Bit 2,7 - Debugging Enable
>This is part of the above Nintendo Logo area, and must be commonly set to 21h, however, Bit 2 and Bit 7 may be set to other values.
>When both bits are set (ie. A5h), the FIQ/Undefined Instruction handler in the BIOS becomes unlocked, the handler then forwards these exceptions to the user handler in cartridge ROM (entry point defined in 80000B4h, see below).
>Other bit combinations currently do not seem to have special functions.

You have that byte hardcoded to have those handlers off, so the device type wouldn't change anything unless the user manually went in and changed that byte (at which point why not change the device type byte as well). Let me know what you think.

**Comment by WinterMute** [comment]
I dug out the gba bios & the abort handlers run through the same handler.

So now I'm thinking we should probably go back to defaulting device type to 0 & using the -d option to set the offset used for the debug handler which would also set the debug enable flag.

Any thoughts?

```
ROM:00000000 ; ---------------------------------------------------------------------------
ROM:00000000 reset
ROM:00000000                 B       reset_handler
ROM:00000004 ; ---------------------------------------------------------------------------
ROM:00000004
ROM:00000004 illegal_instruction
ROM:00000004                 B       fiq_handler
ROM:00000008 ; ---------------------------------------------------------------------------
ROM:00000008
ROM:00000008 swi
ROM:00000008                 B       bios_swi_handler
ROM:0000000C ; ---------------------------------------------------------------------------
ROM:0000000C
ROM:0000000C prefetch_abort
ROM:0000000C                 B       fiq_handler
ROM:00000010 ; ---------------------------------------------------------------------------
ROM:00000010
ROM:00000010 data_abort
ROM:00000010                 B       fiq_handler
ROM:00000014 ; ---------------------------------------------------------------------------
ROM:00000014
ROM:00000014 reserved
ROM:00000014                 B       fiq_handler
ROM:00000018 ; ---------------------------------------------------------------------------
ROM:00000018
ROM:00000018 irq
ROM:00000018                 B       bios_irq_handler
ROM:0000001C ; ---------------------------------------------------------------------------
ROM:0000001C fiq_handler
ROM:0000001C                 LDR     SP, =0x3007FF0
ROM:00000020                 STMFD   SP!, {R12,LR}
ROM:00000024                 MRS     R12, SPSR
ROM:00000028                 MRS     LR, CPSR
ROM:0000002C                 STMFD   SP!, {R12,LR}
ROM:00000030                 MOV     R12, #0x8000000
ROM:00000034                 LDRB    LR, [R12,#(debug_enable_flag - 0x8000000)]
ROM:00000038                 CMP     LR, #0xA5
ROM:0000003C                 BNE     fiq_exit
ROM:00000040                 LDREQB  LR, [R12,#(debug_offset_flag - 0x8000000)]
ROM:00000044                 ANDEQS  LR, LR, #0x80
ROM:00000048                 ADR     LR, fiq_exit
ROM:0000004C                 LDRNE   PC, =debug_handler_9FE2000
ROM:00000050                 LDREQ   PC, =debug_handler_9FFC000
ROM:00000054 fiq_exit
ROM:00000054                 LDR     SP, =0x3007FE0
ROM:00000058                 LDMFD   SP!, {R12,LR}
ROM:0000005C                 MSR     SPSR_cf, R12
ROM:00000060                 LDMFD   SP!, {R12,LR}
ROM:00000064                 SUBS    PC, LR, #4
ROM:00000068 ; ---------------------------------------------------------------------------
```

**Comment by Sierraffinity** [comment]
Yeah that sounds good. I've pushed a commit that adds this change.
